### PR TITLE
IOS-9550 - loading text replaced with spinner

### PR DESCRIPTION
**Pull request by adempsey**
**Risk:** low
**Value:** low, but provides some polish

**Test Cases:**
- Spinner displayed when loading contents correctly
- Spinner disappears when folder contents are displayed
- Spinner is replaced by error message for empty / invalid folders
- Folder contents are replaced by error message if contents are deleted during refresh
- Spinner is not displayed when cached contents are displayed immediately

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @adempsey has signed the CLA. Thanks for the pull request!

File: BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m [hunk 1]
```diff
@@ -419,17 +425,22 @@ - (void)fetchItemsWithCompletion:(void (^)(NSArray *items))completion
...
+    NSString *errorMessage = [[NSString alloc] init];
     if (error == nil) {
-        ((UILabel *)self.tableView.backgroundView).text = NSLocalizedString(@"This folder is empty.", @"Label: Label displayed when the current folder is empty");
+        errorMessage = NSLocalizedString(@"This folder is empty.", @"Label: Label displayed when the current folder is empty");
     } else {
-        ((UILabel *)self.tableView.backgroundView).text = NSLocalizedString(@"Unable to load contents of folder.", @"Label: Label displayed when the current folder is empty");
+        errorMessage = NSLocalizedString(@"Unable to load contents of folder.", @"Label: Label displayed when the current folder is empty");
     }
+    self.folderInfoLabel.text = errorMessage;
+    self.tableView.backgroundView = self.folderInfoLabel;
```
Comment (boxtcarpel) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: set the background view every time we switch to empty state?
Why not just keep the single setting of it in viewDidLoad?

Comment (boxtcarpel) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: Oh, you're using the spinner as the background...

**Comment by boxtcarpel** [comment]
+1

**Comment by boxtcarpel** [comment]
Instead of swapping the background view, what about using a UIView, to which we add a spinner and label as subviews, then alternate which is hidden?

**Comment by adempsey** [comment]
@boxtcarpel see revisions

File: BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m [hunk 2]
```diff
@@ -90,16 +94,34 @@ - (void)viewDidLoad
...
+    self.folderInfoLabel.textColor = [UIColor colorWithRed:180.0f/255.0f green:179.0f/255.0f blue:180.0f/255.0f alpha:1.0f];
+    self.folderInfoLabel.backgroundColor = [UIColor clearColor];
+    self.folderInfoLabel.contentMode = UIViewContentModeCenter;
+    self.folderInfoLabel.numberOfLines = 0;
+    self.folderInfoLabel.text = @"";
+    self.folderInfoLabel.hidden = YES;
+    self.folderInfoLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
+}
+
+- (void)viewWillAppear:(BOOL)animated
```
Comment (RicoYao) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: Should we be setting frames in viewDidLayoutSubviews?

Comment (adempsey) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: The label/activity indicator is misplaced when I set the frames in layout subviews instead:

viewWillAppear:
screen shot 2015-09-17 at 11 23 29 am

viewDidLayoutSubviews:
screen shot 2015-09-17 at 11 22 37 am

Comment (adempsey) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: (accidentally put the pictures backwards, viewDidLayoutSubviews is the off-center label)

Comment (adempsey) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: you were right, it did affect rotation. I moved it to viewDidLayoutSubviews and just fixed the layout.

File: BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m [hunk 3]
```diff
@@ -90,16 +94,35 @@ - (void)viewDidLoad
...
-    emptyLabel.contentMode = UIViewContentModeCenter;
-    emptyLabel.numberOfLines = 0;
-    emptyLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
-    emptyLabel.text = NSLocalizedString(@"Loading…", @"Loading message shown while content is being fetched.");
-    self.tableView.backgroundView = emptyLabel;
+    self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
+    self.activityIndicator.hidesWhenStopped = YES;
+    [self.activityIndicator startAnimating];
+
+    self.folderInfoLabel = [[UILabel alloc] init];
```
Comment (irvingruan) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: Minor: we should move the initialization code into a separate method.

File: BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m [hunk 4]
```diff
@@ -90,16 +94,35 @@ - (void)viewDidLoad
...
+    self.folderInfoLabel.text = @"";
+    self.folderInfoLabel.hidden = YES;
+    self.folderInfoLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
+}
+
+- (void)viewDidLayoutSubviews
+{
+    [super viewDidLayoutSubviews];
+
+    self.backgroundView = [[UIView alloc] initWithFrame:self.view.bounds];
```
Comment (irvingruan) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: Minor: we could move subview-specific layout code into separate methods so they can be triggered elsewhere, if needed.

File: BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m [hunk 5]
```diff
@@ -90,16 +94,35 @@ - (void)viewDidLoad
...
+    self.folderInfoLabel.textColor = [UIColor colorWithRed:180.0f/255.0f green:179.0f/255.0f blue:180.0f/255.0f alpha:1.0f];
+    self.folderInfoLabel.backgroundColor = [UIColor clearColor];
+    self.folderInfoLabel.contentMode = UIViewContentModeCenter;
+    self.folderInfoLabel.numberOfLines = 0;
+    self.folderInfoLabel.text = @"";
+    self.folderInfoLabel.hidden = YES;
+    self.folderInfoLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
+}
+
+- (void)viewDidLayoutSubviews
```
Comment (RicoYao) [inline BoxBrowseSDK/BoxBrowseSDK/BOXFolderViewController.m]: I think we want to do the alloc-ing of views and 'addSubView' calls in viewDidLoad. And then, just set frames here.

**Comment by adempsey** [comment]
@RicoYao see revisions

**Comment by RicoYao** [comment]
+2 lgtm

**Pull request merged**